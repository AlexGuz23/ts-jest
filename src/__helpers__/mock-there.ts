export default function mockThese(map: string[] | { [k: string]: () => any }) {
  const isArray = Array.isArray(map);
  const items: string[] = isArray ? (map as string[]) : Object.keys(map);
  items.forEach(item => {
    const val = isArray ? () => item : map[item];
    jest.doMock(item, val, { virtual: true });
  });
}
