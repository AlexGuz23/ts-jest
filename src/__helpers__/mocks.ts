import { rootLogger } from '../util/logger'
import { testing } from 'bs-logger'

// typings helper
export function mocked<T>(
  val: T,
): T extends (...args: any[]) => any ? jest.MockInstance<T> : jest.Mocked<T> {
  return val as any
}
export function spied<T>(
  val: T,
): T extends (...args: any[]) => any ? jest.SpyInstance<T> : jest.Mocked<T> {
  return val as any
}

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target
