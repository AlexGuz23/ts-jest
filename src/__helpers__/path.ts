import { resolve, relative, join } from 'path'
import { tmpdir } from 'os'
import { ensureDirSync } from 'fs-extra'

export const ROOT = resolve(__dirname, '..', '..')

export function relativeToRoot(path: string) {
  return relative(ROOT, path)
}

export function tempDir(ns: string) {
  const dir = join(
    tmpdir(),
    `ts-jest-unit-tests`,
    ns,
    parseInt((Math.random() * 1e17) as any, 10).toString(36),
  )
  ensureDirSync(dir)
  return dir
}
