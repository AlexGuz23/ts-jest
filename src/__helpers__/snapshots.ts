import outdent from 'outdent'

export const snap = outdent({
  trimTrailingNewline: false,
  trimLeadingNewline: false,
})
