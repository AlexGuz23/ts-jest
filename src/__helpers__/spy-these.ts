export default function spyThese<T extends object, K extends keyof T>(
  object: T,
  implementations: { [key in K]: T[K] | any | undefined },
): { [key in K]: jest.SpyInstance<T[K]> } & { mockRestore: () => void } {
  const keys = Object.keys(implementations) as K[];
  const res = keys.reduce(
    (map, key) => {
      const actual = object[key] as any;
      const spy = jest.spyOn(object, key as K);
      if (implementations[key]) {
        const impl = implementations[key] as (...args: any[]) => any;
        if (impl.length && /\W\$super\W/.test(impl.toString())) {
          spy.mockImplementation(function(this: T, ...args: any[]) {
            return impl.call(this, () => actual.apply(this, args), ...args);
          });
        } else {
          spy.mockImplementation(impl);
        }
      }
      return map;
    },
    {} as any,
  );
  // utility to restore all
  res.mockRestore = () => {
    keys.forEach(key => res[key].mockRestore());
  };
  return res;
}
