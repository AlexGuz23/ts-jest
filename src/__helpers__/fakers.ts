import { TsJestGlobalOptions, BabelConfig } from '../types';
import { resolve } from 'path';
import { ImportReasons } from '../utils/messages';

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath);
}

export const rootDir = filePath('');

export function transpiledTsSource() {
  return `
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var upper_1 = __importDefault(require("./upper"));
var lower_1 = __importDefault(require("./lower"));
jest.mock('./upper', function () { return function (s) { return s.toUpperCase(); }; });
describe('hello', function () {
    test('my test', function () {
        expect(upper_1.default('hello')).toBe('HELLO');
        expect(lower_1.default('HELLO')).toBe('hello');
        jest.mock('./lower', function () { return function (s) { return s.toLowerCase(); }; });
    });
});
`;
}

export function htmlSource() {
  return `
<div>
  <span>some text with \`backtick\`</span>
</div>
`;
}

export function typescriptSource() {
  return `
import upper from './upper';
import lower from './lower';

jest.mock('./upper', () => (s) => s.toUpperCase());

describe('hello', () => {
  test('my test', () => {
    expect(upper('hello')).toBe('HELLO');
    expect(lower('HELLO')).toBe('hello');
    jest.mock('./lower', () => (s) => s.toLowerCase());
  });
});
`;
}

export function tsJestConfig<T extends TsJestGlobalOptions>(
  options?: TsJestGlobalOptions,
): T {
  return { ...options } as any;
}

export function jestConfig<T extends jest.ProjectConfig>(
  options?: jest.InitialOptions,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    moduleFileExtensions: ['ts', 'js'],
    ...options,
  } as any;
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestConfig(tsJestOptions);
  }
  return res;
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any;
}

export function importReason(text: string = 'because'): ImportReasons {
  return text as any;
}
