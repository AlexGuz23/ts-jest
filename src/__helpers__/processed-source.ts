import { RawSourceMap } from 'source-map'
import {
  rewriteSourceMaps,
  relativisePaths,
  parseSource,
  ParsedSourceWithMaps,
} from './source-maps'
import { ROOT } from './path'
import { isAbsolute, relative } from 'path'

// tslint:disable-next-line:no-default-export
export default class ProcessedSource {
  readonly filename: string

  constructor(
    readonly output: string,
    filename: string,
    readonly cwd: string = ROOT,
  ) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)
    return
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap =>
      relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
