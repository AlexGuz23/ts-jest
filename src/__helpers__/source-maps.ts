// WARNING: this file is shared between e2e and unit tests
import { RawSourceMap } from 'source-map'
import bufferFrom from 'buffer-from'
import { relative, isAbsolute, resolve } from 'path'
import stableStringify = require('fast-json-stable-stringify')
import { realpathSync } from 'fs'

// source-map module doesn't provide a sync method to extract the source-maps
export function extractSourceMaps(source: string): RawSourceMap | undefined {
  const [, comment]: [any, string | undefined] =
    (source.match(
      /[\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,(\S+)\s*$/,
    ) as any) || []
  if (!comment) return

  return base64ToSourceMaps(comment)
}

export function base64ToSourceMaps(base64: string): RawSourceMap {
  return JSON.parse(bufferFrom(base64, 'base64').toString('utf8'))
}

export function sourceMapsToBase64(sourceMaps: RawSourceMap): string {
  return bufferFrom(stableStringify(sourceMaps)).toString('base64')
}

export function relativiseSourceRoot(
  fromPath: string,
  source: string,
  prefix: string = '',
): string {
  const from = realpathSync(fromPath)
  const remap = (path: string): string =>
    (isAbsolute(path)
      ? `${prefix}${relative(from, realpathSync(path))}`
      : path
    ).replace(/\\/g, '/')

  return source.replace(
    /([\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,)(\S+)(\s*)$/,
    (_, before, base64, after) => {
      const map = base64ToSourceMaps(base64)
      if (map.sourceRoot) map.sourceRoot = remap(map.sourceRoot)
      if (map.sources) map.sources = map.sources.map(remap)
      if (map.file) map.file = remap(map.file)
      return `${before}${sourceMapsToBase64(map)}${after}`
    },
  )
}
