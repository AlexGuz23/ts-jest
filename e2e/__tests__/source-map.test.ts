import { configureTestCase } from '../__helpers__/test-case'
import { allPackageSets } from '../__helpers__/templates'

const testCase = configureTestCase('source-maps', {
  writeIo: true,
})

testCase.runWithTemplates(allPackageSets, 1, (runTest, { templateName }) => {
  describe(`using template "${templateName}"`, () => {
    const result = runTest()
    it(`should report correct line numbers`, () => {
      expect(result.status).toBe(1)
      expect(result).toMatchSnapshot()
    })

    // when there are some issues with debugging, it's usually becasue source mpas are not inlined
    // and the debugger cannot find the line where to go
    it(`should have the source mpas comment`, () => {
      const io = result.ioFor('main.ts')
      expect(io.normalizedOutputCode).toMatchSnapshot()
      // we could keep only the snapshot testing, but this ensure fixed stuff are the same over versions
      expect(io.outputSourceMaps).toMatchObject({
        names: [],
        sourceRoot: result.cwd,
        sources: ['main.ts'],
        sourcesContent: [io.inputCode],
        version: 3,
      })
    })
  })
})
