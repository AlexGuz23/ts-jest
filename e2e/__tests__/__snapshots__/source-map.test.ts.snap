// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`using template "default" should have the source mpas comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoibWFpbi50cyIsIm1hcHBpbmdzIjoiOztBQUFBLFNBQWdCLFVBQVU7SUFDeEIsT0FBTyxDQUFDLEdBQUcsQ0FBQyxlQUFlLENBQUMsQ0FBQTtBQUM5QixDQUFDO0FBRkQsZ0NBRUM7QUFFRCxTQUFnQixVQUFVO0lBQ3hCLE1BQU0sSUFBSSxLQUFLLENBQUMsZUFBZSxDQUFDLENBQUE7QUFDbEMsQ0FBQztBQUZELGdDQUVDIiwibmFtZXMiOltdLCJzb3VyY2VSb290IjoiPGN3ZD4vIiwic291cmNlcyI6WyJtYWluLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBmdW5jdGlvbiBjb25zb2xlTG9nKCkge1xuICBjb25zb2xlLmxvZygnV0lUSElOIFNPVVJDRScpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiB0aHJvd0Vycm9yKCkge1xuICB0aHJvdyBuZXcgRXJyb3IoJ1dJVEhJTiBTT1VSQ0UnKVxufVxuIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: main.ts
  mappings: >-
    ;;AAAA,SAAgB,UAAU;IACxB,OAAO,CAAC,GAAG,CAAC,eAAe,CAAC,CAAA;AAC9B,CAAC;AAFD,gCAEC;AAED,SAAgB,UAAU;IACxB,MAAM,IAAI,KAAK,CAAC,eAAe,CAAC,CAAA;AAClC,CAAC;AAFD,gCAEC
  names: []
  sourceRoot: <cwd>/
  sources:
    - main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "default" should report correct line numbers 1`] = `
  × jest --runInBand
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-babel-6" should have the source mpas comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJtYXBwaW5ncyI6Ijs7O0FBQUEsU0FBZ0IsVUFBaEIsR0FBMEI7QUFDeEIsWUFBUSxHQUFSLENBQVksZUFBWjtBQUNEO0FBRkQsUUFBQSxVQUFBLEdBQUEsVUFBQTtBQUlBLFNBQWdCLFVBQWhCLEdBQTBCO0FBQ3hCLFVBQU0sSUFBSSxLQUFKLENBQVUsZUFBVixDQUFOO0FBQ0Q7QUFGRCxRQUFBLFVBQUEsR0FBQSxVQUFBIiwibmFtZXMiOltdLCJzb3VyY2VSb290IjoiPGN3ZD4vIiwic291cmNlcyI6WyJtYWluLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBmdW5jdGlvbiBjb25zb2xlTG9nKCkge1xuICBjb25zb2xlLmxvZygnV0lUSElOIFNPVVJDRScpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiB0aHJvd0Vycm9yKCkge1xuICB0aHJvdyBuZXcgRXJyb3IoJ1dJVEhJTiBTT1VSQ0UnKVxufVxuIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  mappings: >-
    ;;;AAAA,SAAgB,UAAhB,GAA0B;AACxB,YAAQ,GAAR,CAAY,eAAZ;AACD;AAFD,QAAA,UAAA,GAAA,UAAA;AAIA,SAAgB,UAAhB,GAA0B;AACxB,UAAM,IAAI,KAAJ,CAAU,eAAV,CAAN;AACD;AAFD,QAAA,UAAA,GAAA,UAAA
  names: []
  sourceRoot: <cwd>/
  sources:
    - main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-babel-6" should report correct line numbers 1`] = `
  × jest --runInBand
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-babel-7" should have the source mpas comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  
  function consoleLog() {
    console.log('WITHIN SOURCE');
  }
  
  exports.consoleLog = consoleLog;
  
  function throwError() {
    throw new Error('WITHIN SOURCE');
  }
  
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJtYXBwaW5ncyI6Ijs7Ozs7O0FBQUEsU0FBZ0IsVUFBaEIsR0FBMEI7QUFDeEIsRUFBQSxPQUFPLENBQUMsR0FBUixDQUFZLGVBQVo7QUFDRDs7QUFGRCxPQUFBLENBQUEsVUFBQSxHQUFBLFVBQUE7O0FBSUEsU0FBZ0IsVUFBaEIsR0FBMEI7QUFDeEIsUUFBTSxJQUFJLEtBQUosQ0FBVSxlQUFWLENBQU47QUFDRDs7QUFGRCxPQUFBLENBQUEsVUFBQSxHQUFBLFVBQUEiLCJuYW1lcyI6W10sInNvdXJjZVJvb3QiOiI8Y3dkPi8iLCJzb3VyY2VzIjpbIm1haW4udHMiXSwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0IGZ1bmN0aW9uIGNvbnNvbGVMb2coKSB7XG4gIGNvbnNvbGUubG9nKCdXSVRISU4gU09VUkNFJylcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIHRocm93RXJyb3IoKSB7XG4gIHRocm93IG5ldyBFcnJvcignV0lUSElOIFNPVVJDRScpXG59XG4iXSwidmVyc2lvbiI6M30=
  ===[ INLINE SOURCE MAPS ]=======================================================
  mappings: >-
    ;;;;;;AAAA,SAAgB,UAAhB,GAA0B;AACxB,EAAA,OAAO,CAAC,GAAR,CAAY,eAAZ;AACD;;AAFD,OAAA,CAAA,UAAA,GAAA,UAAA;;AAIA,SAAgB,UAAhB,GAA0B;AACxB,QAAM,IAAI,KAAJ,CAAU,eAAV,CAAN;AACD;;AAFD,OAAA,CAAA,UAAA,GAAA,UAAA
  names: []
  sourceRoot: <cwd>/
  sources:
    - main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-babel-7" should report correct line numbers 1`] = `
  × jest --runInBand
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-jest-22" should have the source mpas comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoibWFpbi50cyIsIm1hcHBpbmdzIjoiOztBQUFBLFNBQWdCLFVBQVU7SUFDeEIsT0FBTyxDQUFDLEdBQUcsQ0FBQyxlQUFlLENBQUMsQ0FBQTtBQUM5QixDQUFDO0FBRkQsZ0NBRUM7QUFFRCxTQUFnQixVQUFVO0lBQ3hCLE1BQU0sSUFBSSxLQUFLLENBQUMsZUFBZSxDQUFDLENBQUE7QUFDbEMsQ0FBQztBQUZELGdDQUVDIiwibmFtZXMiOltdLCJzb3VyY2VSb290IjoiPGN3ZD4vIiwic291cmNlcyI6WyJtYWluLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBmdW5jdGlvbiBjb25zb2xlTG9nKCkge1xuICBjb25zb2xlLmxvZygnV0lUSElOIFNPVVJDRScpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiB0aHJvd0Vycm9yKCkge1xuICB0aHJvdyBuZXcgRXJyb3IoJ1dJVEhJTiBTT1VSQ0UnKVxufVxuIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: main.ts
  mappings: >-
    ;;AAAA,SAAgB,UAAU;IACxB,OAAO,CAAC,GAAG,CAAC,eAAe,CAAC,CAAA;AAC9B,CAAC;AAFD,gCAEC;AAED,SAAgB,UAAU;IACxB,MAAM,IAAI,KAAK,CAAC,eAAe,CAAC,CAAA;AAClC,CAAC;AAFD,gCAEC
  names: []
  sourceRoot: <cwd>/
  sources:
    - main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-jest-22" should report correct line numbers 1`] = `
  × jest --runInBand
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
        7 | }
        8 | 
        
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
        
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-typescript-2-7" should have the source mpas comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoibWFpbi50cyIsIm1hcHBpbmdzIjoiOztBQUFBO0lBQ0UsT0FBTyxDQUFDLEdBQUcsQ0FBQyxlQUFlLENBQUMsQ0FBQTtBQUM5QixDQUFDO0FBRkQsZ0NBRUM7QUFFRDtJQUNFLE1BQU0sSUFBSSxLQUFLLENBQUMsZUFBZSxDQUFDLENBQUE7QUFDbEMsQ0FBQztBQUZELGdDQUVDIiwibmFtZXMiOltdLCJzb3VyY2VSb290IjoiPGN3ZD4vIiwic291cmNlcyI6WyJtYWluLnRzIl0sInNvdXJjZXNDb250ZW50IjpbImV4cG9ydCBmdW5jdGlvbiBjb25zb2xlTG9nKCkge1xuICBjb25zb2xlLmxvZygnV0lUSElOIFNPVVJDRScpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiB0aHJvd0Vycm9yKCkge1xuICB0aHJvdyBuZXcgRXJyb3IoJ1dJVEhJTiBTT1VSQ0UnKVxufVxuIl0sInZlcnNpb24iOjN9
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: main.ts
  mappings: >-
    ;;AAAA;IACE,OAAO,CAAC,GAAG,CAAC,eAAe,CAAC,CAAA;AAC9B,CAAC;AAFD,gCAEC;AAED;IACE,MAAM,IAAI,KAAK,CAAC,eAAe,CAAC,CAAA;AAClC,CAAC;AAFD,gCAEC
  names: []
  sourceRoot: <cwd>/
  sources:
    - main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-typescript-2-7" should report correct line numbers 1`] = `
  × jest --runInBand
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;
